### DISCUSSION TITLE
Provide tests to verify XML documents before ingestion

### DISCUSSION THREAD

**Discussion by jacobwegner** [scaife-viewer/scaife-viewer.org#30] [Ideas]
Refs https://github.com/scaife-viewer/scaife-viewer.org/discussions/10

Some Scaife Viewer projects have made use of [HookTest](https://github.com/Capitains/HookTest) to test XML, but it seems like there are sometimes false negative produced when checking tests.

I'd like to use this discussion to capture the kinds of tests that I've found useful when ingesting texts during my time working on Scaife Viewer projects; some of the tests may overlap with HookTest, and some may not.

- Test that XML is well-formed
- Test for presence of / validation of teiHeader elements, e.g. for [tlg0059.tlg002.perseus-grc2.](https://github.com/PerseusDL/canonical-greekLit/blob/593087513cb16dd02f0a7b8362519a3a0e2f29bc/data/tlg0059/tlg002/tlg0059.tlg002.perseus-grc2.xml)
  - respStmt
  - publicationStmt
  - refsDecl
- Test of `<text>` children, e.g.
  -  First `div` defines CTS Version kind, has version URN and lang attr:
```xml
<div type="edition" n="urn:cts:greekLit:tlg0059.tlg002.perseus-grc2" xml:lang="grc">
```
- Child(ren) with presence of `@type="textpart"` \*:
```xml
<div type="textpart" subtype="section" resp="perseus" n="17">
```
- Test that a passage can resolve "top-level" textparts from the provided refsDecl (https://scaife.perseus.org/library/urn:cts:greekLit:tlg0059.tlg002.perseus-grc2/cts-api-xml/reffs/)
- Test that a passage can resolve text parts at all levels (a more intensive version of the previous item
- Test that CTS versions provided in `__xml__.py` have corresponding XML files within the work directory

\* We may want to revisit the use of `type=textpart` so that we have a single source of truth between that selector and the `refsDecl` elements; see also https://github.com/scaife-viewer/scaife-viewer/issues/563

**Comment by jacobwegner** [comment]
(I am hoping just to gather some requirements here; I have some prior art for the kinds of tests I'm thinking of, and I think we should probably add them to the `scaife` cli so that we could run them independent of a particular site)

**Comment by pletcher** [comment]
The false negatives that I've noticed often have to do with having the correct elements (`refsDecl`, for example) but with incorrect attributes or children. HookTest doesn't do a great job of explaining why something failed, so using it for validation can sometimes lead to a tedious loop of guessing at a change, running HookTest again, and seeing if it passed -- then rinsing and repeating until HookTest says the file is valid.

But then, as the discussion at https://github.com/PerseusDL/canonical-greekLit/pull/1459 shows, a passing HookTest likely has some false positives as well. (Leaving aside the user error in that PR -- fixing some of those now.)
